lib/AFP/AbstractFilePath.hs

```diff
@@ -0,0 +1,723 @@
+  , addExtension
+  , hasExtension
+  , (<.>)
+  , splitExtensions
+  , dropExtensions
+  , takeExtensions
+  , splitExtension
+
+  -- * Filename\/directory functions
+  , splitFileName
+  , takeFileName
+  , replaceFileName
+  , dropFileName
+  , takeBaseName
+  , replaceBaseName
+  , takeDirectory
+  , replaceDirectory
+  , combine
+  , (</>)
+  , splitPath
+  , joinPath
+  , splitDirectories
+  , takeAllParents
+
+  -- * Drive functions
+  , splitDrive
+  , joinDrive
+  , takeDrive
+  , hasDrive
+  , dropDrive
+  , isDrive
+
+  -- * Trailing slash functions
+  , hasTrailingPathSeparator
+  , addTrailingPathSeparator
+  , dropTrailingPathSeparator
+
+  -- * File name manipulations
+  , normalise
+  , makeRelative
+  , equalFilePath
+  , isRelative
+  , isAbsolute
+  , isValid
+  , makeValid
+  , isFileName
+  , hasParentDir
+  )
+where
+
+import AFP.AbstractFilePath.Internal
+    ( afp
+    , bsToAFP
+    , fromAbstractFilePath
+    , fromAbstractFilePathIO
+    , toAbstractFilePath
+    , toAbstractFilePathIO
+    , unpackAFP
+    , packAFP
+    )
+import AFP.AbstractFilePath.Internal.Types
+    ( AbstractFilePath, PosixFilePath, WindowsFilePath )
+import AFP.OsString
+import AFP.OsString.Internal ( fromChar )
+import AFP.OsString.Internal.Types
+
+#if defined(mingw32_HOST_OS) || defined(__MINGW32__)
+import qualified AFP.AbstractFilePath.Windows as AFP
+#else
+import qualified AFP.AbstractFilePath.Posix as AFP
+#endif
+
+import Data.Bifunctor
+    ( bimap, first )
+
+
+------------------------
+-- Separator predicates
+
+
+-- | Ideal path separator character
+pathSeparator :: OsWord
+pathSeparator = OsWord AFP.pathSeparator
+
+-- | All path separator characters
+pathSeparators :: [OsWord]
+pathSeparators = OsWord <$> AFP.pathSeparators
+
+-- | Check if a character is the path separator
+--
+-- > (n == '/') == isPathSeparator n
+isPathSeparator :: OsWord -> Bool
+isPathSeparator (OsWord w) = AFP.isPathSeparator w
+
+-- | Search path separator
+searchPathSeparator :: OsWord
+searchPathSeparator = OsWord AFP.searchPathSeparator
+
+-- | Check if a character is the search path separator
+--
+-- > (n == ':') == isSearchPathSeparator n
+isSearchPathSeparator :: OsWord -> Bool
+isSearchPathSeparator (OsWord fp) = AFP.isSearchPathSeparator fp
+
+
+-- | File extension separator
+extSeparator :: OsWord
+extSeparator = OsWord AFP.extSeparator
+
+-- | Check if a character is the file extension separator
+--
+-- > (n == '.') == isExtSeparator n
+isExtSeparator :: OsWord -> Bool
+isExtSeparator (OsWord fp) = AFP.isExtSeparator fp
+
+
+------------------------
+-- $PATH methods
+
+
+-- | Take an 'OsString', split it on the 'searchPathSeparator'.
+-- On Windows path elements are stripped of quotes.
+-- On Posix blank items are converted to @.@.
+--
+-- Follows the recommendations in
+-- <http://www.opengroup.org/onlinepubs/009695399/basedefs/xbd_chap08.html>
+--
+-- > Windows: splitSearchPath "File1;File2;File3" == ["File1","File2","File3"]
```

Comment:
This is slightly unfortunate. In this module I want to document both `Windows` and `Posix`, while in the specific posix/windows modules we may have diverging doctests. So there's some duplication, that may lead to documentation-rot.